### ISSUE TITLE
t.p.syslog should support facilities and levels.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3300]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| TimAllen reported|
|-|-|
|Trac ID|trac#3300|
|Type|enhancement|
|Created|2008-06-20 07:19:18Z|
|Branch|https://github.com/twisted/twisted/tree/syslog-levels-3300|

As described on [the mailing list](http://twistedmatrix.com/pipermail/twisted-python/2008-June/017932.html), I'm interested in extending twistd's syslog support. Glyph [encouraged me to submit my suggestions as patches](http://twistedmatrix.com/pipermail/twisted-python/2008-June/017941.html), so I'll attach a patch-series to this bug. These patches should apply cleanly to r23947.

Attachments:

* [0001-Add-docstrings-to-twisted.python.syslog.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/512cfbebb1245a9dcbe11e23a7e245e9e9d5c6e9.patch) (1144 bytes) - added by TimAllen on 2008-06-20 07:19:59Z - Patch 1/5: Add docstrings to twisted.python.syslog
* [0002-Basic-support-for-syslog-options-and-facility-suppor.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/3b291c9e70a144ca5f0d9d02e453681debd54ab3.patch) (2290 bytes) - added by TimAllen on 2008-06-20 07:20:35Z - Basic support for syslog options and facilities.
* [0003-SyslogObserver-should-interpret-message-contents-to.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/78034106648a34e0dbf7880ec92b3aa0b9b7dc8c.patch) (5683 bytes) - added by TimAllen on 2008-06-20 07:21:34Z - SyslogObserver should set priority and facility based on message contents.
* [0004-Add-a-handy-method-to-look-up-facilities-by-name.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/5ebc8d6c19d66b39e3e7ce8ea4ccf371ccb5f783.patch) (2333 bytes) - added by TimAllen on 2008-06-20 07:22:04Z - Add a handy method to look up facilities by name.
* [0005-Teach-twistd-about-configuring-syslog-options-and-fa.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/4654835211f107aa1f7b351afad72594b5368d80.patch) (5109 bytes) - added by TimAllen on 2008-06-20 07:22:34Z - Teach twistd about configuring syslog options and facilities.
* [0002-t.p.syslog.Syslog-should-call-t.p.l.textFromEventDic.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/a5de52f4f9d7755b8848e9b6305798671590a357.patch) (921 bytes) - added by TimAllen on 2008-06-27 06:11:27Z - syslog.SyslogObserver.emit() should call t.p.l.textFromEventDict(), not copy it.
* [0003-Basic-support-for-syslog-options-and-facility-suppor.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/884765a0fcf6426554c0160cba9e6e34105b10d8.patch) (2302 bytes) - added by TimAllen on 2008-06-27 06:14:58Z - Basic support for syslog options and facilities.
* [0004-SyslogObserver-should-interpret-message-contents-to.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/929df1cb10bd1d4ae9dda36ad4a7683430f430b2.patch) (5369 bytes) - added by TimAllen on 2008-06-27 06:15:55Z - {{{SyslogObserver}}} should set priority and facility based on message contents.
* [0005-Add-a-handy-method-to-look-up-facilities-by-name.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/53a0d7ef5c743b99116be0a95110a93dbedd16da.patch) (2333 bytes) - added by TimAllen on 2008-06-27 06:16:50Z - Add a handy method to look up facilities by name.
* [0006-Teach-twistd-about-configuring-syslog-options-and-fa.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/4313cfc7a17e30f06589305cd36f0d1e166bae4a.patch) (5109 bytes) - added by TimAllen on 2008-06-27 06:17:19Z - Teach twistd about configuring syslog options and facilities.
* [twistd-syslog-3300.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d56/d566dce13cebefa2c17e16faf2d602be94b4e50d/9d4d42719617abbff1593e740295bf08e7f90c61.patch) (11184 bytes) - added by TimAllen on 2009-09-15 06:58:28Z - A patch that just upgrades Twisted's syslog support.

<details><summary>Searchable metadata</summary>

```
trac-id__3300 3300
type__enhancement enhancement
reporter__TimAllen TimAllen
priority__normal normal
milestone__ 
branch__branches_syslog_levels_3300 branches/syslog-levels-3300
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1213946358000000 1213946358000000
changetime__1253173917000000 1253173917000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|TimAllen commented|
|-|-|

These patches are intended to be applied in order, and "```trial twisted```" should pass cleanly after each patch.

 * ["0001-Add-docstrings-to-twisted.python.syslog.patch"](None/"0001-Add-docstrings-to-twisted.python.syslog.patch") just adds some docstrings for tidiness.
 * ["0002-Basic-support-for-syslog-options-and-facility-suppor.patch"](None/"0002-Basic-support-for-syslog-options-and-facility-suppor.patch") is the minimum syslog support I would need in Twisted.
 * ["0003-SyslogObserver-should-interpret-message-contents-to.patch"](None/"0003-SyslogObserver-should-interpret-message-contents-to.patch") extends the ```SyslogObserver``` object to interpret fields in the ```eventDict``` and set appropriate syslog flags. Specifically, any message including a ```syslogPriority``` will be logged at that priority; any message including a ```syslogFacility``` will be logged to that facility; and any message that ```isError``` will be logged at ```LOG_ALERT``` priority unless overridden.
 * By this point, ```twisted.python.syslog``` is about as comprehensive as I can imagine any twisted logging <-> syslog connector being.
 * ["0004-Add-a-handy-method-to-look-up-facilities-by-name.patch"](None/"0004-Add-a-handy-method-to-look-up-facilities-by-name.patch") adds a utility function to ```twisted.python.syslog``` that will be used in the next patch.
 * ["0005-Teach-twistd-about-configuring-syslog-options-and-fa.patch"](None/"0005-Teach-twistd-about-configuring-syslog-options-and-fa.patch") is the one that actually hooks this new functionality into twistd. twistd doesn't have a whole lot of unit tests, but I started up the 'web' plugin with "```--syslogFacility=LOG_LOCAL0```" and observed that logging went to the right place.

For cross reference purposes, other bugs mentioned in the original mailing-list thread include [#11359](https://github.com/twisted/twisted/issues/11359) and [#2571](https://github.com/twisted/twisted/issues/2571).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|TimAllen commented|
|-|-|

It's just occurred to me that `SyslogObserver.emit` should probably be calling `t.p.log.textFromEventDict` at some point, rather than duplicating its logic.

I'll try to get around to that shortly.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|TimAllen removed owner|
|-|-|

An updated patch sequence, with a fix for comment:3.

Cleanup patches:
 * ["0001-Add-docstrings-to-twisted.python.syslog.patch"](None/"0001-Add-docstrings-to-twisted.python.syslog.patch") just adds some docstrings for tidiness. 
 * ["0002-t.p.syslog.Syslog-should-call-t.p.l.textFromEventDic.patch"](None/"0002-t.p.syslog.Syslog-should-call-t.p.l.textFromEventDic.patch") makes ```t.p.syslog.SyslogObserver.emit()``` call ```t.p.l.textEventFromDict()```. It looks like it was originally copied from there anyway.

Completing the syslog functionality:
 * ["0003-Basic-support-for-syslog-options-and-facility-suppor.patch"](None/"0003-Basic-support-for-syslog-options-and-facility-suppor.patch") is the minimum syslog support I would need in Twisted. 
 * ["0004-SyslogObserver-should-interpret-message-contents-to.patch"](None/"0004-SyslogObserver-should-interpret-message-contents-to.patch") extends the ```SyslogObserver``` object to interpret fields in the ```eventDict``` and set appropriate syslog flags. Specifically, any message including a ```syslogPriority``` will be logged at that priority; any message including a ```syslogFacility``` will be logged to that facility; and any message that ```isError``` will be logged at ```LOG_ALERT``` priority unless overridden.

```twistd``` integration patches:
 * ["0005-Add-a-handy-method-to-look-up-facilities-by-name.patch"](None/"0005-Add-a-handy-method-to-look-up-facilities-by-name.patch") adds a utility function to twisted.python.syslog that will be used in the next patch.
 * ["0006-Teach-twistd-about-configuring-syslog-options-and-fa.patch"](None/"0006-Teach-twistd-about-configuring-syslog-options-and-fa.patch") is the one that actually hooks this new functionality into twistd. twistd doesn't have a whole lot of unit tests, but I started up the 'web' plugin with "```--syslogFacility=LOG_LOCAL0```" and observed that logging went to the right place.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [24181]) Branching to 'twistd-syslog-3300'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [24182]) Apply 0001-Add-docstrings-to-twisted.python.syslog.patch

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [24183]) Apply 0002-t.p.syslog.Syslog-should-call-t.p.l.textFromEventDic.patch 

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [24184]) Apply 0003-Basic-support-for-syslog-options-and-facility-suppor.patch 

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [24185]) Apply 0004-SyslogObserver-should-interpret-message-contents-to.patch 

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [24186]) Apply 0005-Add-a-handy-method-to-look-up-facilities-by-name.patch 

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [24187]) Apply 0006-Teach-twistd-about-configuring-syslog-options-and-fa.patch 

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for submitting this!  I'm trying something new, which is reviewing and responding to the review more or less simultaneously.  I don't know if this is a good idea, we'll see.

  * twisted/python/syslog.py
    * missing a test-case-name
    * copyright date needs to be updated
    * the existing `import log` should be replaced with an absolute import
    * `STR_TO_FACILITY` would be shorter and simpler if it were just a `getattr(syslog, name)`
    * there should be three blank lines between top-level suites (ie, function definitions) and two between suites in a class definition
    * if an expression has to be broken inside parenthesis, as the parameter list to `SyslogObserver.__init__` is, the next line should be indented in one of two ways:
      * if the newline on the first line comes immediately after the opening (, the next line should be indented 4 spaces.
      * otherwise, the next line should be indented to line up with the first column after the (.
    * arguments to `__init__` which do not end up as attributes of the same name should be documented in the docstring of `__init__`.
    * `SyslogObserver.emit` is missing a docstring
    * interpolation should always use a tuple (or a dict) as the right-hand argument, even for just one value being interpolated.  In the exception message in `facilityFromString`, the facility name is interpolated all by itself.  If someone does something really dumb, like pass in `('foo', 'bar')`, this will result in a `TypeError`.
  * twisted/scripts/_twistd_unix.py
    * `syslogOptions`, `syslogFacility` should be `syslog-options`, `syslog-facility`.
    * `syslogOptions` is a bit unfriendly :)  bitwise-or isn't a nice thing to ask people to enter on the command line.  Probably this should be `--syslog-option` with multiple applications allowed and accepting a string like NOWAIT.
    * the syslog options do need unit tests
    * the syslogOptions entry in `optParameters` specifies a default for that key, and then `UnixAppLogger` also specifies a default.  One of these is redundant.  The same goes for the syslogFacility key.
    * `SyslogObserver.emit` needs unit tests
  * twisted/test/test_twistd.py
    * a number of tests use integer literals with comments about symbolic names which correspond to those literals - they should just use the symbolic names

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [24190]) address review comments

refs [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @exarkun|
|-|-|

OK, I have to say that I am -1 for any modification of twistd. The enhancements of syslog are nice, but the configuration of the logger should happen with [#11359](https://github.com/twisted/twisted/issues/11359), not with yet other options. So here are my suggestions:
 * revert all changes to twistd
 * move all the tests for `SyslogObserver` in twisted.python.test.test_syslog (why where they in test_twistd ?)
 * I suspect some care are needed for skipping under windows

Thanks.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|TimAllen removed owner|
|-|-|

OK, a shiny new version of this patch. This version is based on exarkun's version in comment:13, addressing some of the issues raised by therve in comment:16. Specifically:

 * This patch does not touch `twistd`, although it does touch `twistd`'s tests because they made assumptions about how `t.p.syslog` was importing the stdlib syslog.
 * I have moved the `t.p.syslog` tests to `t.p.test.test_syslog`.
 * I have changed the way the tests import syslog so that they should be quietly skipped on platforms without syslog support.
 * I added a test and a docstring that had been added to my employer's internal copy of the code after the original patches had been attached to this bug.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve set status to `closed`|
|-|-|

(In [27349]) Merge syslog-levels-3300

Authors: TimAllen, exarkun
Reviewers: exarkun, therve
Fixes [#3300](https://github.com/twisted/twisted/issues/3300)

Add support for syslog levels and facilities in the Twisted syslog observer.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set status to `reopened`|
|-|-|

(In [27351]) Revert r27349: commit contained tabs.

Reopens [#3300](https://github.com/twisted/twisted/issues/3300)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve set status to `closed`|
|-|-|

(In [27352]) Merge syslog-levels-3300

Authors: TimAllen, exarkun
Reviewers: exarkun, therve
Fixes [#3300](https://github.com/twisted/twisted/issues/3300)

Add support for syslog levels and facilities in the Twisted syslog observer.  This is retry
from r27349 because it contained tabs (thanks Amaury!).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
